File: Virtual Machine/Nova.VM/JumpIfFalseCode.cpp
#include "JumpIfFalseCode.h"

void JumpIfFalseCode::Compute(RuntimeContext& context,RuntimeContext::RuntimeElement* locales, int& index)
{
	int value = std::get<int>(context.PopStack()); // value must be char.

	if (value == 0)
	{
		index = this->TargetIndex;
	}
	else
	{
		index++;
	}
}

void JumpIfFalseCode::Deserialize(BinaryReader& reader)
{
	this->TargetIndex = reader.Read<int>();
}

File: Compiler/Nova.Bytecode/Codes/JumpIfFalseCode.cs
using Nova.ByteCode.Runtime;
using Nova.Utils.IO;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nova.ByteCode.Codes
{
    public class JumpIfFalseCode : ICode
    {
        public int TypeId => 4;

        public int targetIndex;

        public JumpIfFalseCode(int index)
        {
            this.targetIndex = index;
        }

        public void Compute(RuntimeContext context, ref object[] locals, ref int index)
        {
            if ((int)context.PopStack() == 0)
            {
                index = this.targetIndex;
            }
            else
            {
                index++;
            }
        }
        public override string ToString()
        {
            return "(" + TypeId + ") " +  "JumpIfFalse " + targetIndex;
        }

        public void Serialize(CppBinaryWriter writer)
        {
            writer.Write(targetIndex);
        }
    }
}
